activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb

```diff
@@ -1510,6 +1530,25 @@ def valid_primary_key_options # :nodoc:
       end
 
       private
+        MAX_GENERATED_INDEX_NAME_BYTES = 62
+
+        def generate_index_name(table_name, column)
+          name = "index_#{table_name}_on_#{Array(column) * '_and_'}"
+          return name if name.bytesize <= MAX_GENERATED_INDEX_NAME_BYTES
+
+          # Remove _and_ between columns
+          name = "index_#{table_name}_on_#{Array(column) * '_'}"
+          return name if name.bytesize <= MAX_GENERATED_INDEX_NAME_BYTES
```

Comment:
I wonder if this "semi-fallback" is really worth it. The behavior may be more easy to understand by users if we limit ourselves to just two formats.